STARLARK.md

```diff
@@ -0,0 +1,547 @@
+
+Simple example with default values and complete documentation:
+
+```Python
+def ec2_instance_builder(
+        image=RE_WINDOWS_JDK7_IMAGE,
+        platform=RE_WINDOWS_PLATFORM,
+        instance_type=RE_LARGE_INSTANCE_TYPE,
+        preemptible=False,
+        use_ssd=True
+):
+    """
+    Base configuration for a VM that uses a pre-defined image provided by RE Team.
+
+    Provides the default values for the VM configuration:
+    - image: the image to use for the VM by default RE_JAVA_17_IMAGE
+    - platform: the platform of the VM, by default "windows"
+    - instance_type: the type of the VM, by default RE_LARGE_INSTANCE_TYPE
+    - preemptible: whether the VM is preemptible, by default False
+    - use_ssd: whether the VM uses SSD, by default True
+
+    :return: a dictionary with the configuration for the VM
+    """
+    return {
+        "experimental": "true",
+        "image": image,
+        "platform": platform,
+        "region": RE_DEFAULT_REGION,
+        "type": instance_type,
+        "subnet_id": "${CIRRUS_AWS_SUBNET}",
+        "preemptible": preemptible,
+        "use_ssd": use_ssd,
+    }
+```
+
+Advanced example without control flow and with a lot of configuration options:
+
+```Python
+def build_task_builder(task_id, name, alias, depends_on, **kwargs):
+    task_conf = {
+        "name": name,
+        "alias": alias,
+        "eks_container": default_container_builder(True),
+        "depends_on": depends_on,
+        "set_docker_host_script": set_docker_host_script(),
+    }
+    task_conf |= gradle_cache()
+    task_conf |= kwargs
+    return {task_id: task_conf}
+```
+
+### Recursive aggregation of configuration
+
+Cirrus looks at the file .cirrus.star in order to generate the pipeline. It’s very easy to turn this file into a mess by
+importing dozens of modules and calling tons of functions from it.
+
+In general when a script or a class suffers from too many responsibilities it’s a good idea to split it into smaller
+chunks.
+
+One way to get a cleaner code is to consider the Starlark code base as a tree of modules. Each module is a node that
+provides one unique consolidated configuration to its ancestor. Each module implements a `<module domain>_all_tasks`
+function returning all the configurations for the domain. Then `.cirrus.star` does not have to know about how this is
+implemented and is not exposed to the changes in the implementation.
+
+This is also closely related to the “Separation of concern” principle.
+
+This pattern can also be applied to a shared module.
+
+`.cirrus.star` content:
+
+```Python
+load("github.com/SonarSource/cirrus-modules@v2", "load_features")
+load(".cirrus/modules/base.star", "base_all_tasks")
+load(".cirrus/modules/build.star", "build_all_tasks")
+load(".cirrus/modules/deploy.star", "deploy_all_tasks")
+load(".cirrus/modules/qa.star", "qa_all_tasks")
+load(".cirrus/modules/housekeeping.star", "housekeeping_all_tasks")
+
+
+def main(ctx):
+    conf = {}
```

Comment:
In code this line is defined as `conf = dict()`. This should be consistent.

Reply:
Indeed!
Good catch!